icicle-compiler/test/Icicle/Test/Foreign/Array.hs

```diff
@@ -0,0 +1,271 @@
+
+instance Arbitrary Input where
+  arbitrary = genInput
+
+opaqueRet :: Ty -> RetType WordPtr
+opaqueRet ty = case ty of
+  IDouble
+    -> ptrToWordPtr <$> retPtr retDouble
+  IInt
+    -> ptrToWordPtr <$> retPtr retInt64
+  IError
+    -> ptrToWordPtr <$> retPtr retInt64
+  ITime
+    -> ptrToWordPtr <$> retPtr retString
+  IUnit
+    -> ptrToWordPtr <$> retPtr retVoid
+  IBool
+    -> ptrToWordPtr <$> retPtr retInt64
+  IString
+    -> ptrToWordPtr <$> retPtr retString
+  IArray t
+    -> ptrToWordPtr <$> (wordPtrToPtr <$> opaqueRet t)
+
+genValForType :: Ty -> Gen Val
+genValForType ty = case ty of
+  IDouble
+    -> VDouble <$> arbitrary
+  IInt
+    -> VInt <$> arbitrary
+  IBool
+    -> VBool <$> arbitrary
+  IError
+    -> return $ VError Icicle.ExceptTombstone
+  ITime
+    -> VTime <$> arbitrary
+  IString
+    -> VString <$> elements simpsons
+  IUnit
+    -> return VUnit
+  IArray t
+    -> VArray <$> listOf (genValForType t)
+
+genInput :: Gen Input
+genInput = do
+  ty   <- arbitrary
+  val  <- listOf (genValForType ty)
+  return $ Input ty val
+
+seaTestFunName :: Text -> Text
+seaTestFunName fn
+   = "test_" <> fn
+
+seaOfType :: Ty -> Text
+seaOfType ty = case ty of
+  IDouble  -> "idouble"
+  IInt     -> "iint"
+  IError   -> "ierror"
+  IBool    -> "ibool"
+  ITime    -> "itime"
+  IString  -> "istring"
+  IUnit    -> "iunit"
+  IArray t -> "iarray_" <> seaOfType t
+
+runWith :: Input -> (WordPtr -> IO Property) -> IO Property
+runWith (Input _ val) prop = go val
+  where
+    go xs = do
+      (array, _free) <- pokeArr xs
+      r <- prop array
+      -- Let the caller free the allocated pointers as they will go through the ffi.
+      -- _free
+      return r
+
+pokeVal :: Val -> IO (WordPtr, IO ())
+pokeVal v = case v of
+  VInt x
+    -> pokeOne x
+  VDouble x
+    -> pokeOne x
+  VError _
+    -> pokeOne ()
+  VBool x
+    -> pokeOne x
+  VTime x
+    -> pokeOne (packedOfTime x)
+  VUnit
+    -> pokeOne ()
+  VString xs
+    -> pokeStr xs
+  VArray xs
+    -> pokeArr xs
+
+pokeOne x = do
+  ptr <- malloc
+  pokeElemOff ptr 0 x
+  return (ptrToWordPtr ptr, free ptr)
+
+pokeStr xs = do
+  ptr <- callocBytes (length xs + 1)
+  zipWithM_ (pokeElemOff ptr) [0..] xs
+  return (ptrToWordPtr ptr, free ptr)
+
+pokeArr xs = do
+  (ps, fs) <- List.unzip <$> mapM pokeVal xs
+  ptr <- mallocBytes (length xs + 1)
+  pokeByteOff ptr 0 (length xs)
+  zipWithM_ (pokeElemOff ptr) [1..] ps
+  return . (ptrToWordPtr ptr,) $ foldr (>>) (return ()) fs
+
+--------------------------------------------------------------------------------
+
+runRight :: (Monad m, Show a) => EitherT a m Property -> m Property
+runRight a = do
+  e <- runEitherT a
+  case e of
+    Left  x -> return (counterexample (show x) failed)
+    Right x -> return x
+
+retBool :: RetType Bool
+retBool = mkStorableRetType ffi_type_uint64
+
+
+return []
+tests :: IO Bool
+tests = $forAllProperties $ quickCheckWithResult (stdArgs { maxSuccess = 100, maxSize = 100 })
```

Comment:
This probable needs `releaseLibraryAfterTests`, see Icicle.Test.Data.Time for an example